File: C/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    if (!stack) return NULL;
    stack->head = NULL;
    stack->size = 0;
    return stack;
}

int getSize(Stack* stack) {
    return stack ? stack->size : 0;
}

bool isEmpty(Stack* stack) {
    return stack->head == NULL;
}

void push(Stack* stack, void* data) {
    if (!stack) return;

    Node* new_node = createNode(data);
    if (!new_node) return;

    new_node->next = stack->head;
    stack->head = new_node;
    stack->size++;
}

void* pop(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Error: La cola esta vacia\n");
        return NULL;
    }

    Node* temp = stack->head;
    void* data = temp->data;
    stack->head = temp->next;
    free(temp);
    stack->size--;
    return data;
}

void* peek(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Error: La cola esta vacia\n");
        return NULL;
    }
    return stack->head->data;
}

void showStack(Stack* stack, void (*print_func)(void*)) {
    if (isEmpty(stack)) {
        printf("La cola esta vacia.\n");
        return;
    }

    Node* current = stack->head;
    while (current) {
        print_func(current->data);
        printf(" -> ");
        current = current->next;
    }
    printf("NULL\n");
}

void freeStack(Stack* stack) {
    while (!isEmpty(stack)) {
        pop(stack);
    }
    free(stack);
}

File: Java/Stack.java
public class Stack<T> {
	private Node<T> head;
	private int size;
	
	public Stack() {
		this.head = null;
		this.size = 0;
	}
	
	public int getSize() {
		return size;
	}
	
	public boolean isEmpty() {
		return head == null;
	}
	
	public void push(T data) {
		Node<T> newNode = new Node<>(data);
		newNode.setNext(head);
		head = newNode;
		size++;
	}
	
	public T pop() {
		if (isEmpty()) {
			throw new IllegalStateException("La pila esta vacia");
		}
		
		T data = head.getData();
		head = head.getNext();
		size--;
		return data;
	}
	
	public T peek() {
		if (isEmpty()) {
			throw new IllegalStateException("La pila esta vacia");
		}
		
		return head.getData();
	}
	
	public void showStack() {
		Node<T> currentNode = head;
		
		while (currentNode != null) {
			System.out.print(currentNode.getData() + " -> ");
			currentNode = currentNode.getNext();
		}
		
		System.out.println("null");
	}
}
